projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,154 @@
+// interrupt is raised, prv_interrupt_handler reads the conversion values from the registers and
+// stores them in storage. The user could access this data by the provided read functions. These
+// functions get their data from storage and not directly by reading the registers. Also look at
+// ads1015_def for more information on macros and how registers are setup.
+#include <status.h>
+#include <stdio.h>
+#include "ads1015_def.h"
+#include "gpio_it.h"
+
+// reset the internal registers and enter a power-down state
+static StatusCode prv_reset(Ads1015Storage *storage) {
+  uint8_t reset = ADS1015_RESET_BYTE;
+  return i2c_write(storage->i2c_port, ADS1015_I2C_GENERAL_CALL, &reset, sizeof(reset));
+}
+
+// writes to register given upper and lower bytes
+static StatusCode prv_setup_register(Ads1015Storage *storage, uint8_t reg, uint8_t msb,
+                                     uint8_t lsb) {
+  uint8_t ads1015_setup_register[] = { reg, msb, lsb };
+  return i2c_write(storage->i2c_port, storage->i2c_addr + ADS1015_I2C_BASE_ADDRESS,
+                   ads1015_setup_register, SIZEOF_ARRAY(ads1015_setup_register));
+}
+
+// reads the register and stores the value in the given array
+static StatusCode prv_read_register(I2CPort i2c_port, Ads1015Address i2c_addr, uint8_t reg,
+                                    uint8_t *read, size_t size_read) {
+  status_ok_or_return(i2c_write(i2c_port, i2c_addr + ADS1015_I2C_BASE_ADDRESS, &reg, sizeof(reg)));
+
+  status_ok_or_return(i2c_read(i2c_port, i2c_addr + ADS1015_I2C_BASE_ADDRESS, read, size_read));
+  return STATUS_CODE_OK;
+}
+
+// Switches to the the given channel by writing to config reg
+static StatusCode prv_set_channel(Ads1015Storage *storage, Ads1015Channel channel) {
+  if (channel >= NUM_ADS1015_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_CONFIG,
+                                         CONFIG_REGISTER_MSB(channel), CONFIG_REGISTER_LSB));
+  storage->current_channel = channel;
+  storage->channel_enable[channel] = true;
+  return STATUS_CODE_OK;
+}
+
+// This function is registered as the callback for ALRT/RDY Pin.
+// reads and stores the conversion value in storage, and switches to the next enabled channel.
+static void prv_interrupt_handler(const GPIOAddress *address, void *context) {
+  Ads1015Storage *storage = context;
+
+  uint8_t read_conv_register[2];
+  prv_read_register(storage->i2c_port, storage->i2c_addr, ADS1015_ADDRESS_POINTER_CONV,
+                    read_conv_register, SIZEOF_ARRAY(read_conv_register));
+
+  uint8_t config_register[2];
+  prv_read_register(storage->i2c_port, storage->i2c_addr, ADS1015_ADDRESS_POINTER_CONFIG,
+                    config_register, SIZEOF_ARRAY(config_register));
+
+  Ads1015Channel current_channel = storage->current_channel;
+  // following line puts the two read bytes into an int16.
+  storage->channel_readings[current_channel] =
```

Comment:
You should probably just sign-extend here to be consistent. Otherwise, your stored channel reading is technically off by a factor.